pocketutils/misc/klgists.py

```diff
@@ -0,0 +1,829 @@
+	std_level = logging.INFO,
+	file_level = logging.DEBUG
+):
+	"""Initializes a logger that can write to a log file and/or stdout."""
+
+	if log_path is not None and not os.path.exists(os.path.dirname(log_path)):
+		os.mkdir(os.path.dirname(log_path))
+
+	if child_logger_name is None:
+		logger = logging.getLogger()
+	else:
+		logger = logging.getLogger(child_logger_name)
+	logger.setLevel(logging.NOTSET)
+
+	formatter = logging.Formatter(format_str)
+
+	if log_path is not None:
+		handler = logging.FileHandler(log_path, encoding='utf-8')
+		handler.setLevel(file_level)
+		handler.setFormatter(formatter)
+		logger.addHandler(handler)
+
+	if to_std:
+		handler = logging.StreamHandler()
+		handler.setLevel(std_level)
+		handler.setFormatter(formatter)
+		logger.addHandler(handler)
+
+import datetime
+def format_time(time: datetime) -> str:
+	"""Standard timestamp format. Ex: 2016-05-02_22_35_56."""
+	return time.strftime("%Y-%m-%d_%H-%M-%S")
+
+def timestamp() -> str:
+	"""Standard timestamp of time now. Ex: 2016-05-02_22_35_56."""
+	return format_time(datetime.datetime.now())
+
+def timestamp_path(path: str) -> str:
+	"""Standard way to label a file path with a timestamp."""
+	return "{}-{}".format(path, timestamp())
+
+
+def nice_time(n_ms: int) -> str:
+	length = datetime.datetime(1, 1, 1) + datetime.timedelta(milliseconds=n_ms)
+	if n_ms < 1000 * 60 * 60 * 24:
+		return "{}h, {}m, {}s".format(length.hour, length.minute, length.second)
+	else:
+		return "{}d, {}h, {}m, {}s".format(length.day, length.hour, length.minute, length.second)
+
+
+def parse_local_iso_datetime(z: str) -> datetime:
+	return datetime.datetime.strptime(z, '%Y-%m-%dT%H:%M:%S.%f')
+
+
+logger = logging.getLogger(__name__)
+
+class PipeType(Enum):
+	STDOUT = 1
+	STDERR = 2
+
+def _disp(out, ell, name):
+	out = out.strip()
+	if '\n' in out:
+		ell(name + ":\n<<=====\n" + out + '\n=====>>')
+	elif len(out) > 0:
+		ell(name + ": <<===== " + out + " =====>>")
+	else:
+		ell(name + ": <no output>")
+	
+
+def _log(out, err, ell):
+	_disp(out, ell, "stdout")
+	_disp(err, ell, "stderr")
+
+	
+def smart_log_callback(source, line, prefix: str = '') -> None:
+	line = line.decode('utf-8')
+	if line.startswith('FATAL:'):
+		logger.fatal(prefix + line)
+	elif line.startswith('ERROR:'):
+		logger.error(prefix + line)
+	elif line.startswith('WARNING:'):
+		logger.warning(prefix + line)
+	elif line.startswith('INFO:'):
+		logger.info(prefix + line)
+	elif line.startswith('DEBUG:'):
+		logger.debug(prefix + line)
+	else:
+		logger.debug(prefix + line)
+
+
+def _reader(pipe_type, pipe, queue):
+	try:
+		with pipe:
+			for line in iter(pipe.readline, b''):
+				queue.put((pipe_type, line))
+	finally:
+		queue.put(None)
+	
+def stream_cmd_call(cmd: List[str], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell_cmd: str=None, cwd: Optional[str] = None, timeout_secs: Optional[float] = None, log_callback: Callable[[PipeType, bytes], None] = None, bufsize: Optional[int] = None) -> None:
```

Comment:
This is [CallTools `steram_cmd_call`](https://github.com/dmyersturnbull/pocketutils/blob/main/pocketutils/tools/call_tools.py#L115).